rero_ils/modules/contributions/sync.py

```diff
@@ -0,0 +1,546 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019 RERO
+# Copyright (C) 2020 UCLouvain
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""API for manipulating documents."""
+
+import logging.config
+import os
+import sys
+from copy import deepcopy
+from datetime import datetime
+from itertools import islice
+
+import requests
+from deepdiff import DeepDiff
+from elasticsearch_dsl import Q
+from flask import current_app
+
+from rero_ils.modules.contributions.api import Contribution, \
+    ContributionsSearch
+from rero_ils.modules.documents.api import Document, DocumentsSearch
+from rero_ils.modules.utils import get_timestamp, set_timestamp
+
+
+class SyncAgent(object):
+    """Agent MEF synchronization."""
+
+    def __init__(self, dry_run=False, verbose=False, log_dir=None,
+                 from_last_date=False):
+        """Constructor.
+
+        :param dry_run: bool - if true the data are not modified
+        :param verbose: bool or integer - verbose level
+        :param log_dir: string - path to put the logs
+        :param from_last_date: boolean - if True try to consider agent modified
+            after the last run date time
+        """
+        self.dry_run = dry_run
+        self.verbose = verbose
+        self.from_date = None
+        self.start_timestamp = None
+        self._init_logger(verbose, log_dir)
+        if from_last_date:
+            self._get_last_date()
+
+    def _get_last_date(self):
+        """Get the date of the last execution of the synchronization."""
+        self.from_date = None
+        data = get_timestamp('sync_agents')
+        if data and data.get('start_timestamp'):
+            self.from_date = data.get('start_timestamp')
+
+    def _init_logger(self, verbose, log_dir):
+        """Initialize the module logger."""
+        # default value
+        if not log_dir:
+            log_dir = current_app.config.get(
+                'RERO_ILS_MEF_SYNC_LOG_DIR',
+                os.path.join(current_app.instance_path, 'logs')
+            )
+        # create the log directory if does not exists
+        if not os.path.exists(log_dir):
+            os.mkdir(log_dir)
+        verbose_level = ['ERROR', 'INFO', 'DEBUG']
+        logging_config = {
```

Comment:
maybe it should be part of a `conf/logging.py` file into this module ? just a suggestion